File: app/src/main/java/com/example/pam_app/adapter/BucketListAdapter.java
package com.example.pam_app.adapter;

import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import com.example.pam_app.R;
import com.example.pam_app.utils.listener.ClickableTarget;
import com.example.pam_app.model.Bucket;

import java.util.ArrayList;
import java.util.List;

public class BucketListAdapter extends RecyclerView.Adapter<BucketListAdapter.BucketViewHolder> {
    private final List<Bucket> bucketList;
    private ClickableTarget listener;

    public BucketListAdapter() {
        this.bucketList = new ArrayList<>();
    }

    public void setOnClickListener(final ClickableTarget listener) {
        this.listener = listener;
    }

    public void update(final List<Bucket> newBucketList) {
        bucketList.clear();
        if (newBucketList != null) {
            bucketList.addAll(newBucketList);
            notifyDataSetChanged();
        }
    }

    public void showNewBucket(final Bucket bucket) {
        if (bucket != null) {
            bucketList.add(bucket);
            notifyDataSetChanged();
        }
    }

    @NonNull
    @Override
    public BucketViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        final View view = LayoutInflater.from(parent.getContext()).inflate(R.layout.bucket_list_item, parent, false);
        return new BucketViewHolder(view);
    }

    @Override
    public void onBindViewHolder(@NonNull BucketViewHolder holder, int position) {
        holder.bind(bucketList.get(position));
        holder.setOnClickListener(listener);
    }

    @Override
    public int getItemCount() {
        return bucketList.size();
    }

    public static class BucketViewHolder extends RecyclerView.ViewHolder {
        private ClickableTarget listener;

        public BucketViewHolder(@NonNull View itemView) {
            super(itemView);
        }

        public void bind(Bucket bucket) {
            final TextView title = itemView.findViewById(R.id.bucket_title);
            final TextView target = itemView.findViewById(R.id.bucket_target);

            title.setText(bucket.title);
            target.setText(new StringBuilder().append("$").append(bucket.target));

            itemView.setOnClickListener(v -> {
                if (listener != null) {
                    listener.onClick(bucket.id);
                }
            });
        }

        public void setOnClickListener(ClickableTarget listener) {
            this.listener = listener;
        }
    }
}

File: app/src/main/java/com/example/pam_app/adapter/BucketListAdapter.kt
package com.example.pam_app.adapter

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import androidx.recyclerview.widget.RecyclerView
import com.example.pam_app.R
import com.example.pam_app.adapter.BucketListAdapter.BucketViewHolder
import com.example.pam_app.model.Bucket
import com.example.pam_app.utils.listener.ClickableTarget
import java.util.*
import java.util.stream.Collectors

class BucketListAdapter : RecyclerView.Adapter<BucketViewHolder>() {
    private var bucketList: MutableList<Bucket?>
    private var listener: ClickableTarget<Int>? = null
    fun setOnClickListener(listener: ClickableTarget<Int>?) {
        this.listener = listener
    }

    fun update(newBucketList: List<Bucket?>?) {
        bucketList.clear()
        if (newBucketList != null) {
            bucketList.addAll(newBucketList)
            notifyDataSetChanged()
        }
    }

    fun showNewBucket(bucket: Bucket?) {
        if (bucket != null) {
            bucketList.add(bucket)
            notifyDataSetChanged()
        }
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): BucketViewHolder {
        val view = LayoutInflater.from(parent.context).inflate(R.layout.bucket_list_item, parent, false)
        return BucketViewHolder(view)
    }

    override fun onBindViewHolder(holder: BucketViewHolder, position: Int) {
        holder.bind(bucketList[position])
        holder.setOnClickListener(listener)
    }

    override fun getItemCount(): Int {
        return bucketList.size
    }

    fun delete(id: Int) {
        bucketList = bucketList.stream().filter { b: Bucket? -> b!!.id != id }.collect(Collectors.toList())
        notifyDataSetChanged()
    }

    class BucketViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        private var listener: ClickableTarget<Int>? = null
        fun bind(bucket: Bucket?) {
            val title = itemView.findViewById<TextView>(R.id.bucket_title)
            val target = itemView.findViewById<TextView>(R.id.bucket_target)
            title.text = bucket!!.title
            target.text = StringBuilder().append("$").append(bucket.target)
            itemView.setOnClickListener {
                if (listener != null) {
                    listener!!.onClick(bucket.id!!)
                }
            }
        }

        fun setOnClickListener(listener: ClickableTarget<Int>?) {
            this.listener = listener
        }
    }

    init {
        bucketList = ArrayList()
    }
}